pkg/controller/daemon/manager.go

```diff
@@ -0,0 +1,475 @@
+	eventBroadcaster.StartRecordingToSink(kubeClient.Events(""))
+
+	dm := &DaemonManager{
+		kubeClient: kubeClient,
+		podControl: controller.RealPodControl{
+			KubeClient: kubeClient,
+			Recorder:   eventBroadcaster.NewRecorder(api.EventSource{Component: "daemon"}),
+		},
+		expectations: controller.NewControllerExpectations(),
+		queue:        workqueue.New(),
+	}
+	// Manage addition/update of daemon controllers.
+	dm.dcStore.Store, dm.dcController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Experimental().Daemons(api.NamespaceAll).List(labels.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Experimental().Daemons(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&expapi.Daemon{},
+		FullDaemonResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				daemon := obj.(*expapi.Daemon)
+				glog.V(4).Infof("Adding daemon %s", daemon.Name)
+				dm.enqueueController(obj)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				oldDaemon := old.(*expapi.Daemon)
+				glog.V(4).Infof("Updating daemon %s", oldDaemon.Name)
+				dm.enqueueController(cur)
+			},
+			DeleteFunc: func(obj interface{}) {
+				daemon := obj.(*expapi.Daemon)
+				glog.V(4).Infof("Deleting daemon %s", daemon.Name)
+				dm.enqueueController(obj)
+			},
+		},
+	)
+	// Watch for creation/deletion of pods. The reason we watch is that we don't want a daemon controller to create/delete
+	// more pods until all the effects (expectations) of a daemon controller's create/delete have been observed.
+	dm.podStore.Store, dm.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Pod{},
+		FullDaemonPodResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addPod,
+			UpdateFunc: dm.updatePod,
+			DeleteFunc: dm.deletePod,
+		},
+	)
+	// Watch for new nodes or updates to nodes - daemons are launched on new nodes, and possibly when labels on nodes change,
+	dm.nodeStore.Store, dm.nodeController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Nodes().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Nodes().Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Node{},
+		FullNodeResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addNode,
+			UpdateFunc: dm.updateNode,
+			DeleteFunc: func(node interface{}) {},
```

Comment:
what's expected to happen when nodes die, and why don't we care about it here? isn't there some state to clean up?